File: frontend/src/models/questTemplate.ts
/**
 * Quest Template models for the GoalsGuild application.
 * 
 * This file contains TypeScript interfaces for Quest Template operations
 * including creation, updates, and responses. All interfaces follow
 * the established patterns from the quest models.
 */

import { QuestDifficulty, QuestKind, QuestCountScope } from './quest';

// Quest template privacy options
export type QuestTemplatePrivacy = 'public' | 'followers' | 'private';

// Quest template creation input
export interface QuestTemplateCreateInput {
  title: string;
  description?: string;
  category: string;
  difficulty: QuestDifficulty;
  rewardXp: number;
  tags: string[];
  privacy: QuestTemplatePrivacy;
  kind: QuestKind;
  targetCount?: number;
  countScope?: QuestCountScope;
  estimatedDuration?: number;
  instructions?: string;
}

// Quest template update input
export interface QuestTemplateUpdateInput {
  title?: string;
  description?: string;
  category?: string;
  difficulty?: QuestDifficulty;
  rewardXp?: number;
  tags?: string[];
  privacy?: QuestTemplatePrivacy;
  kind?: QuestKind;
  targetCount?: number;
  countScope?: QuestCountScope;
}

// Quest template response
export interface QuestTemplate {
  id: string;
  userId: string;
  title: string;
  description?: string;
  category: string;
  difficulty: QuestDifficulty;
  rewardXp: number;
  tags: string[];
  privacy: QuestTemplatePrivacy;
  kind: QuestKind;
  targetCount?: number;
  countScope?: QuestCountScope;
  createdAt: number;
  updatedAt: number;
}

// Quest template list response
export interface QuestTemplateListResponse {
  templates: QuestTemplate[];
  total: number;
  hasMore: boolean;
  nextToken?: string;
}

// Quest template creation form data
export interface QuestTemplateFormData {
  title: string;
  description: string;
  category: string;
  difficulty: QuestDifficulty;
  rewardXp: number;
  tags: string[];
  privacy: QuestTemplatePrivacy;
  kind: QuestKind;
  targetCount: number;
  countScope: QuestCountScope;
}

// Quest template validation errors
export interface QuestTemplateValidationErrors {
  [key: string]: string | undefined;
}

// Quest template filter options
export interface QuestTemplateFilters {
  search?: string;
  category?: string;
  difficulty?: QuestDifficulty;
  privacy?: QuestTemplatePrivacy;
  kind?: QuestKind;
}

// Quest template sort options
export interface QuestTemplateSortOptions {
  field: 'title' | 'createdAt' | 'updatedAt' | 'difficulty' | 'rewardXp';
  direction: 'asc' | 'desc';
}

// Quest template list options
export interface QuestTemplateListOptions {
  limit?: number;
  nextToken?: string;
  privacy?: 'user' | 'public' | 'all';
  filters?: QuestTemplateFilters;
  sort?: QuestTemplateSortOptions;
}

// Privacy level descriptions
export const PRIVACY_DESCRIPTIONS: Record<QuestTemplatePrivacy, string> = {
  public: 'Anyone can see and use this template',
  followers: 'Only your followers can see and use this template',
  private: 'Only you can see and use this template'
};

// Privacy level icons
export const PRIVACY_ICONS: Record<QuestTemplatePrivacy, string> = {
  public: '🌍',
  followers: '👥',
  private: '🔒'
};

// Privacy level colors
export const PRIVACY_COLORS: Record<QuestTemplatePrivacy, string> = {
  public: 'text-green-600',
  followers: 'text-blue-600',
  private: 'text-gray-600'
};

// Default quest template form data
export const DEFAULT_QUEST_TEMPLATE_FORM_DATA: QuestTemplateFormData = {
  title: '',
  description: '',
  category: '',
  difficulty: 'medium',
  rewardXp: 50,
  tags: [],
  privacy: 'private',
  kind: 'linked',
  targetCount: 1,
  countScope: 'completed_tasks'
};

// Quest template categories (same as quest categories)
export const QUEST_TEMPLATE_CATEGORIES = [
  'Health', 'Work', 'Personal', 'Learning', 'Fitness', 'Creative', 
  'Financial', 'Social', 'Spiritual', 'Hobby', 'Travel', 'Other'
];

// Quest template difficulty options
export const QUEST_TEMPLATE_DIFFICULTY_OPTIONS: Array<{
  value: QuestDifficulty;
  label: string;
  description: string;
}> = [
  {
    value: 'easy',
    label: 'Easy',
    description: 'Simple tasks that can be completed quickly'
  },
  {
    value: 'medium',
    label: 'Medium',
    description: 'Moderate tasks requiring some effort'
  },
  {
    value: 'hard',
    label: 'Hard',
    description: 'Challenging tasks requiring significant effort'
  }
];

// Quest template privacy options
export const QUEST_TEMPLATE_PRIVACY_OPTIONS: Array<{
  value: QuestTemplatePrivacy;
  label: string;
  description: string;
  icon: string;
}> = [
  {
    value: 'public',
    label: 'Public',
    description: 'Anyone can see and use this template',
    icon: '🌍'
  },
  {
    value: 'followers',
    label: 'Followers',
    description: 'Only your followers can see and use this template',
    icon: '👥'
  },
  {
    value: 'private',
    label: 'Private',
    description: 'Only you can see and use this template',
    icon: '🔒'
  }
];

// Quest template kind options
export const QUEST_TEMPLATE_KIND_OPTIONS: Array<{
  value: QuestKind;
  label: string;
  description: string;
}> = [
  {
    value: 'linked',
    label: 'Linked Quest',
    description: 'Quest based on completing specific tasks or goals'
  },
  {
    value: 'quantitative',
    label: 'Quantitative Quest',
    description: 'Quest based on completing a certain number of items'
  }
];

// Quest template count scope options
export const QUEST_TEMPLATE_COUNT_SCOPE_OPTIONS: Array<{
  value: QuestCountScope;
  label: string;
  description: string;
}> = [
  {
    value: 'completed_tasks',
    label: 'Completed Tasks',
    description: 'Count completed tasks'
  },
  {
    value: 'completed_goals',
    label: 'Completed Goals',
    description: 'Count completed goals'
  }
];

// Utility functions
export const getPrivacyDescription = (privacy: QuestTemplatePrivacy): string => {
  return PRIVACY_DESCRIPTIONS[privacy];
};

export const getPrivacyIcon = (privacy: QuestTemplatePrivacy): string => {
  return PRIVACY_ICONS[privacy];
};

export const getPrivacyColor = (privacy: QuestTemplatePrivacy): string => {
  return PRIVACY_COLORS[privacy];
};

export const formatTemplateTitle = (title: string): string => {
  return title.trim();
};

export const formatTemplateDescription = (description: string): string => {
  return description.trim();
};

export const formatTemplateTags = (tags: string[]): string[] => {
  return tags
    .map(tag => tag.trim())
    .filter(tag => tag.length > 0)
    .slice(0, 10); // Limit to 10 tags
};

export const validateTemplateTitle = (title: string): string | null => {
  if (!title || title.trim().length < 3) {
    return 'Title must be at least 3 characters long';
  }
  if (title.trim().length > 100) {
    return 'Title must be 100 characters or less';
  }
  return null;
};

export const validateTemplateDescription = (description: string): string | null => {
  if (description && description.trim().length > 500) {
    return 'Description must be 500 characters or less';
  }
  return null;
};

export const validateTemplateTags = (tags: string[]): string | null => {
  if (tags.length > 10) {
    return 'Maximum 10 tags allowed';
  }
  for (const tag of tags) {
    if (tag.trim().length > 20) {
      return 'Each tag must be 20 characters or less';
    }
  }
  return null;
};

export const validateTemplateTargetCount = (targetCount: number, kind: QuestKind): string | null => {
  if (kind === 'quantitative') {
    if (!targetCount || targetCount < 1) {
      return 'Target count is required for quantitative quests';
    }
    if (targetCount > 1000) {
      return 'Target count must be 1000 or less';
    }
  }
  return null;
};

export const validateTemplateCountScope = (countScope: QuestCountScope, kind: QuestKind): string | null => {
  if (kind === 'quantitative') {
    if (!countScope) {
      return 'Count scope is required for quantitative quests';
    }
  }
  return null;
};

export const validateQuestTemplate = (template: QuestTemplateFormData): QuestTemplateValidationErrors => {
  const errors: QuestTemplateValidationErrors = {};

  const titleError = validateTemplateTitle(template.title);
  if (titleError) errors.title = titleError;

  const descriptionError = validateTemplateDescription(template.description);
  if (descriptionError) errors.description = descriptionError;

  const tagsError = validateTemplateTags(template.tags);
  if (tagsError) errors.tags = tagsError;

  const targetCountError = validateTemplateTargetCount(template.targetCount, template.kind);
  if (targetCountError) errors.targetCount = targetCountError;

  const countScopeError = validateTemplateCountScope(template.countScope, template.kind);
  if (countScopeError) errors.countScope = countScopeError;

  return errors;
};

export const hasValidationErrors = (errors: QuestTemplateValidationErrors): boolean => {
  return Object.keys(errors).length > 0;
};

File: backend/services/quest-service/app/models/quest_template.py
"""
Quest Template models for the GoalsGuild Quest feature.

This module contains Pydantic models for Quest Template operations including
creation, updates, and responses. All models include comprehensive
validation, XSS protection, and follow the existing quest-service patterns.
"""

from typing import List, Optional, Literal
from pydantic import BaseModel, Field, field_validator
from datetime import datetime
import re

# Import quest-related types from the main quest models
from .quest import QuestDifficulty, QuestKind, QuestCountScope

# Quest template privacy options
QuestTemplatePrivacy = Literal["public", "followers", "private"]

# Validation constants
MAX_TITLE_LENGTH = 100
MIN_TITLE_LENGTH = 3
MAX_DESCRIPTION_LENGTH = 500
MAX_TAGS_COUNT = 10
MAX_TAG_LENGTH = 20
MAX_REWARD_XP = 1000
MIN_REWARD_XP = 0
DEFAULT_REWARD_XP = 50


class QuestTemplateCreatePayload(BaseModel):
    """Payload for creating a new quest template"""
    
    # Basic Information (Required)
    title: str = Field(
        min_length=MIN_TITLE_LENGTH,
        max_length=MAX_TITLE_LENGTH,
        description="Template title"
    )
    description: Optional[str] = Field(
        default=None,
        max_length=MAX_DESCRIPTION_LENGTH,
        description="Template description"
    )
    category: str = Field(
        min_length=1,
        max_length=50,
        description="Template category"
    )
    difficulty: QuestDifficulty = Field(description="Template difficulty level")
    rewardXp: int = Field(
        default=DEFAULT_REWARD_XP,
        ge=MIN_REWARD_XP,
        le=MAX_REWARD_XP,
        description="XP reward for completing quests from this template"
    )
    tags: List[str] = Field(
        default_factory=list,
        max_length=MAX_TAGS_COUNT,
        description="Template tags"
    )
    privacy: QuestTemplatePrivacy = Field(
        default="private",
        description="Template privacy level"
    )
    kind: QuestKind = Field(description="Template quest kind")
    
    # Template-specific fields
    targetCount: Optional[int] = Field(
        default=None,
        ge=1,
        le=1000,
        description="Target count for quantitative quests"
    )
    countScope: Optional[QuestCountScope] = Field(
        default=None,
        description="Count scope for quantitative quests"
    )
    
    @field_validator('title')
    @classmethod
    def validate_title(cls, v: str) -> str:
        """Validate and sanitize title"""
        if not v or not v.strip():
            raise ValueError("Title cannot be empty")
        
        # Basic XSS protection
        sanitized = re.sub(r'<[^>]+>', '', v.strip())
        if len(sanitized) < MIN_TITLE_LENGTH:
            raise ValueError(f"Title must be at least {MIN_TITLE_LENGTH} characters")
        
        return sanitized
    
    @field_validator('description')
    @classmethod
    def validate_description(cls, v: Optional[str]) -> Optional[str]:
        """Validate and sanitize description"""
        if not v:
            return None
        
        # Basic XSS protection
        sanitized = re.sub(r'<[^>]+>', '', v.strip())
        if len(sanitized) > MAX_DESCRIPTION_LENGTH:
            raise ValueError(f"Description must be {MAX_DESCRIPTION_LENGTH} characters or less")
        
        return sanitized if sanitized else None
    
    @field_validator('category')
    @classmethod
    def validate_category(cls, v: str) -> str:
        """Validate category"""
        if not v or not v.strip():
            raise ValueError("Category cannot be empty")
        
        # Basic XSS protection
        sanitized = re.sub(r'<[^>]+>', '', v.strip())
        if len(sanitized) > 50:
            raise ValueError("Category must be 50 characters or less")
        
        return sanitized
    
    @field_validator('tags')
    @classmethod
    def validate_tags(cls, v: List[str]) -> List[str]:
        """Validate and sanitize tags"""
        if len(v) > MAX_TAGS_COUNT:
            raise ValueError(f"Maximum {MAX_TAGS_COUNT} tags allowed")
        
        sanitized_tags = []
        for tag in v:
            if not tag or not tag.strip():
                continue
            
            # Basic XSS protection and validation
            sanitized = re.sub(r'<[^>]+>', '', tag.strip())
            if len(sanitized) > MAX_TAG_LENGTH:
                raise ValueError(f"Tag '{tag}' must be {MAX_TAG_LENGTH} characters or less")
            
            if len(sanitized) < 1:
                continue
                
            sanitized_tags.append(sanitized)
        
        return sanitized_tags
    
    @field_validator('targetCount')
    @classmethod
    def validate_target_count(cls, v: Optional[int], info) -> Optional[int]:
        """Validate target count based on quest kind"""
        if v is None:
            return None
        
        # For quantitative quests, target count is required
        if info.data.get('kind') == 'quantitative' and v is None:
            raise ValueError("Target count is required for quantitative quest templates")
        
        return v
    
    @field_validator('countScope')
    @classmethod
    def validate_count_scope(cls, v: Optional[QuestCountScope], info) -> Optional[QuestCountScope]:
        """Validate count scope based on quest kind"""
        if v is None:
            return None
        
        # For quantitative quests, count scope is required
        if info.data.get('kind') == 'quantitative' and v is None:
            raise ValueError("Count scope is required for quantitative quest templates")
        
        return v


class QuestTemplateUpdatePayload(BaseModel):
    """Payload for updating an existing quest template"""
    
    # All fields are optional for updates
    title: Optional[str] = Field(
        default=None,
        min_length=MIN_TITLE_LENGTH,
        max_length=MAX_TITLE_LENGTH,
        description="Template title"
    )
    description: Optional[str] = Field(
        default=None,
        max_length=MAX_DESCRIPTION_LENGTH,
        description="Template description"
    )
    category: Optional[str] = Field(
        default=None,
        min_length=1,
        max_length=50,
        description="Template category"
    )
    difficulty: Optional[QuestDifficulty] = Field(
        default=None,
        description="Template difficulty level"
    )
    rewardXp: Optional[int] = Field(
        default=None,
        ge=MIN_REWARD_XP,
        le=MAX_REWARD_XP,
        description="XP reward for completing quests from this template"
    )
    tags: Optional[List[str]] = Field(
        default=None,
        max_length=MAX_TAGS_COUNT,
        description="Template tags"
    )
    privacy: Optional[QuestTemplatePrivacy] = Field(
        default=None,
        description="Template privacy level"
    )
    kind: Optional[QuestKind] = Field(
        default=None,
        description="Template quest kind"
    )
    targetCount: Optional[int] = Field(
        default=None,
        ge=1,
        le=1000,
        description="Target count for quantitative quests"
    )
    countScope: Optional[QuestCountScope] = Field(
        default=None,
        description="Count scope for quantitative quests"
    )
    
    # Apply same validators as create payload
    @field_validator('title')
    @classmethod
    def validate_title(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return None
        return QuestTemplateCreatePayload.validate_title(v)
    
    @field_validator('description')
    @classmethod
    def validate_description(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return None
        return QuestTemplateCreatePayload.validate_description(v)
    
    @field_validator('category')
    @classmethod
    def validate_category(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return None
        return QuestTemplateCreatePayload.validate_category(v)
    
    @field_validator('tags')
    @classmethod
    def validate_tags(cls, v: Optional[List[str]]) -> Optional[List[str]]:
        if v is None:
            return None
        return QuestTemplateCreatePayload.validate_tags(v)


class QuestTemplateResponse(BaseModel):
    """Response model for quest template operations"""
    
    id: str = Field(description="Template ID")
    userId: str = Field(description="User ID who created the template")
    title: str = Field(description="Template title")
    description: Optional[str] = Field(description="Template description")
    category: str = Field(description="Template category")
    difficulty: QuestDifficulty = Field(description="Template difficulty level")
    rewardXp: int = Field(description="XP reward for completing quests from this template")
    tags: List[str] = Field(description="Template tags")
    privacy: QuestTemplatePrivacy = Field(description="Template privacy level")
    kind: QuestKind = Field(description="Template quest kind")
    targetCount: Optional[int] = Field(description="Target count for quantitative quests")
    countScope: Optional[QuestCountScope] = Field(description="Count scope for quantitative quests")
    createdAt: int = Field(description="Template creation timestamp")
    updatedAt: int = Field(description="Template last update timestamp")
    
    class Config:
        from_attributes = True


class QuestTemplateListResponse(BaseModel):
    """Response model for listing quest templates"""
    
    templates: List[QuestTemplateResponse] = Field(description="List of quest templates")
    total: int = Field(description="Total number of templates")
    hasMore: bool = Field(description="Whether there are more templates available")
    nextToken: Optional[str] = Field(description="Token for pagination")
